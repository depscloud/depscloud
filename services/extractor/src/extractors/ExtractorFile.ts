import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
