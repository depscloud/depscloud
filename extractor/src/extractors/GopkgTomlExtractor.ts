import {Dependency, DependencyManagementFile} from "@depscloud/api/v1alpha/deps";
import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import inferImportPath from "./goutils/inferImportPath";
import parseImportPath from "./goutils/parseImportPath";
import Languages from "./Languages";
import MatchConfig from "../matcher/MatchConfig";

interface Constraint {
    name: string;
    version: string;
    branch: string;
    revision: string;
}

function transformConstraints(data: Constraint[], scope: string): Dependency[] {
    return (data || []).map(({ name, version, branch, revision }) => {
        const { organization, module } = parseImportPath(name);

        let versionConstraint = version;
        if (branch) {
            versionConstraint = branch;
        } else if (revision) {
            versionConstraint = revision;
        }

        return {
            organization,
            module,
            versionConstraint,
            scopes: [ scope ],
            name,
        };
    });
}

function transformSimple(data: string[], versionConstraint: string, scope: string): Dependency[] {
    return (data || []).map((name) => {
        const { organization, module } = parseImportPath(name);

        return {
            organization,
            module,
            versionConstraint,
            scopes: [ scope ],
            name,
        };
    });
}

export default class GopkgTomlExtractor implements Extractor {
    public matchConfig(): MatchConfig {
        return {
            includes: [
                "**/Gopkg.toml",
            ],
            excludes: [
                "**/vendor/**"
            ],
        };
    }

    public requires(): string[] {
        return [ "Gopkg.toml" ];
    }

    public async extract(url: string, files: { [p: string]: ExtractorFile }): Promise<DependencyManagementFile> {
        const name = inferImportPath(url);
        const { organization, module } = parseImportPath(name);

        const toml = files["Gopkg.toml"].toml();

        const dependencies: Dependency[] = [];
        dependencies.push(...transformConstraints(toml.constraint, "constraint"));
        dependencies.push(...transformConstraints(toml.override, "override"));
        dependencies.push(...transformSimple(toml.ignored, "*", "ignored"));
        dependencies.push(...transformSimple(toml.required, "*", "required"));

        return {
            language: Languages.GO,
            system: "gopkg",
            sourceUrl: "",
            organization,
            module,
            version: "latest",
            dependencies,
            name: name,
        };
    }
}
