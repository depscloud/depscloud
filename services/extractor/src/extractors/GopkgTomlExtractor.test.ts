import {readFile} from "fs";
import {promisify} from "util";
import ExtractorFile from "./ExtractorFile";
import GopkgTomlExtractor from "./GopkgTomlExtractor";

const readFileAsync = promisify(readFile);

describe("GopkgTomlExtractor", () => {
    test("fullParse", async () => {
        const gopkgToml = require.resolve("./testdata/Gopkg.toml");
        const buffer = await readFileAsync(gopkgToml);
        const content = buffer.toString();

        const parser = new GopkgTomlExtractor();

        const actual = await parser.extract(
            "git@github.com:depscloud/extractor.git", {
                "Gopkg.toml": new ExtractorFile(content),
            },
        );

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
