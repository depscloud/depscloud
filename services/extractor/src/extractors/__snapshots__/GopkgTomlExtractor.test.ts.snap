// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GopkgTomlExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "cloud.google.com/go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": ">=0.20.0, <1.0.0",
    },
    Object {
      "name": "go.opencensus.io",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": ">=0.15.0",
    },
    Object {
      "name": "contrib.go.opencensus.io/exporter/stackdriver",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": ">=0.2.0",
    },
    Object {
      "name": "code.cloudfoundry.org/copilot",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/DataDog/datadog-go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.0.0",
    },
    Object {
      "name": "github.com/alicebob/miniredis",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.3.2",
    },
    Object {
      "name": "github.com/aws/aws-sdk-go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.13.24",
    },
    Object {
      "name": "github.com/cactus/go-statsd-client",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "3.1.1",
    },
    Object {
      "name": "github.com/cenkalti/backoff",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.0.0",
    },
    Object {
      "name": "github.com/circonus-labs/circonus-gometrics",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.1.0",
    },
    Object {
      "name": "github.com/coreos/go-oidc",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "v2",
    },
    Object {
      "name": "github.com/davecgh/go-spew",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.1.0",
    },
    Object {
      "name": "github.com/envoyproxy/go-control-plane",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "^0.6.7",
    },
    Object {
      "name": "github.com/fluent/fluent-logger-golang",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.3.0",
    },
    Object {
      "name": "github.com/ghodss/yaml",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "github.com/go-redis/redis",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "6.10.2",
    },
    Object {
      "name": "github.com/gogo/googleapis",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/gogo/protobuf",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "^1.1.1",
    },
    Object {
      "name": "github.com/golang/protobuf",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.1.0",
    },
    Object {
      "name": "github.com/golang/sync",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/google/go-github",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "15.0.0",
    },
    Object {
      "name": "github.com/google/uuid",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.2.0",
    },
    Object {
      "name": "github.com/googleapis/gax-go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.0.0",
    },
    Object {
      "name": "github.com/gorilla/mux",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.6.1",
    },
    Object {
      "name": "github.com/gorilla/websocket",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.2.0",
    },
    Object {
      "name": "github.com/grpc-ecosystem/go-grpc-middleware",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/grpc-ecosystem/go-grpc-prometheus",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.1.0",
    },
    Object {
      "name": "github.com/grpc-ecosystem/grpc-opentracing",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/hashicorp/consul",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.3.0",
    },
    Object {
      "name": "github.com/hashicorp/go-multierror",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/hashicorp/vault",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.10.0",
    },
    Object {
      "name": "github.com/howeyc/fsnotify",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.9.0",
    },
    Object {
      "name": "github.com/natefinch/lumberjack",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.1.0",
    },
    Object {
      "name": "github.com/onsi/gomega",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.3.0",
    },
    Object {
      "name": "github.com/open-policy-agent/opa",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.8.2",
    },
    Object {
      "name": "github.com/openshift/api",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/opentracing/opentracing-go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.0.2",
    },
    Object {
      "name": "github.com/pmezard/go-difflib",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "github.com/prometheus/client_golang",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.9.2",
    },
    Object {
      "name": "github.com/prometheus/client_model",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/prometheus/common",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/prometheus/prom2json",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "github.com/satori/go.uuid",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.2.0",
    },
    Object {
      "name": "github.com/spf13/cobra",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.0.2",
    },
    Object {
      "name": "github.com/stretchr/testify",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.2.1",
    },
    Object {
      "name": "github.com/uber/jaeger-client-go",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "go.uber.org/atomic",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.2.0",
    },
    Object {
      "name": "go.uber.org/multierr",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.1.0",
    },
    Object {
      "name": "go.uber.org/zap",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "1.7.1",
    },
    Object {
      "name": "golang.org/x/net",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "golang.org/x/oauth2",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "golang.org/x/time",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "golang.org/x/tools",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "google.golang.org/api",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "google.golang.org/genproto",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "google.golang.org/grpc",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "^1.13.0",
    },
    Object {
      "name": "gopkg.in/yaml.v2",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.2.1",
    },
    Object {
      "name": "istio.io/api",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "k8s.io/api",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "release-1.11",
    },
    Object {
      "name": "k8s.io/apimachinery",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "release-1.11",
    },
    Object {
      "name": "k8s.io/apiextensions-apiserver",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "release-1.11",
    },
    Object {
      "name": "k8s.io/helm",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "2.9.1",
    },
    Object {
      "name": "k8s.io/kubernetes",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "release-1.11",
    },
    Object {
      "name": "github.com/gogo/status",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "d60b5acac426cb52f7e28124c6aaa862d9f339f5",
    },
    Object {
      "name": "github.com/google/go-cmp",
      "scopes": Array [
        "constraint",
      ],
      "versionConstraint": "0.2.0",
    },
    Object {
      "name": "github.com/docker/distribution",
      "scopes": Array [
        "override",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "k8s.io/client-go",
      "scopes": Array [
        "override",
      ],
      "versionConstraint": "release-8.0",
    },
    Object {
      "name": "github.com/Azure/go-autorest",
      "scopes": Array [
        "override",
      ],
      "versionConstraint": "bca49d5b51a50dc5bb17bbf6204c711c6dbded06",
    },
    Object {
      "name": "github.com/golang/glog",
      "scopes": Array [
        "override",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "name": "github.com/imdario/mergo",
      "scopes": Array [
        "override",
      ],
      "versionConstraint": "6633656539c1639d9d78127b7d47c622b5d7b6dc",
    },
    Object {
      "name": "istio.io/istio/mixer/adapter/stackdriver/contextgraph",
      "scopes": Array [
        "ignored",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "github.com/client9/misspell",
      "scopes": Array [
        "required",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "github.com/istio/tools/protoc-gen-docs",
      "scopes": Array [
        "required",
      ],
      "versionConstraint": "*",
    },
    Object {
      "name": "fortio.org/fortio",
      "scopes": Array [
        "required",
      ],
      "versionConstraint": "*",
    },
  ],
  "language": "go",
  "name": "github.com/depscloud/extractor",
  "sourceUrl": "",
  "system": "gopkg",
  "version": "latest",
}
`;

exports[`GopkgTomlExtractor fullParse 2`] = `
"{
  \\"language\\": \\"go\\",
  \\"system\\": \\"gopkg\\",
  \\"sourceUrl\\": \\"\\",
  \\"name\\": \\"github.com/depscloud/extractor\\",
  \\"version\\": \\"latest\\",
  \\"dependencies\\": [
    {
      \\"name\\": \\"cloud.google.com/go\\",
      \\"versionConstraint\\": \\">=0.20.0, <1.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"go.opencensus.io\\",
      \\"versionConstraint\\": \\">=0.15.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"contrib.go.opencensus.io/exporter/stackdriver\\",
      \\"versionConstraint\\": \\">=0.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"code.cloudfoundry.org/copilot\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/DataDog/datadog-go\\",
      \\"versionConstraint\\": \\"2.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/alicebob/miniredis\\",
      \\"versionConstraint\\": \\"2.3.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/aws/aws-sdk-go\\",
      \\"versionConstraint\\": \\"1.13.24\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/cactus/go-statsd-client\\",
      \\"versionConstraint\\": \\"3.1.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/cenkalti/backoff\\",
      \\"versionConstraint\\": \\"2.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/circonus-labs/circonus-gometrics\\",
      \\"versionConstraint\\": \\"2.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/coreos/go-oidc\\",
      \\"versionConstraint\\": \\"v2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/davecgh/go-spew\\",
      \\"versionConstraint\\": \\"1.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/envoyproxy/go-control-plane\\",
      \\"versionConstraint\\": \\"^0.6.7\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/fluent/fluent-logger-golang\\",
      \\"versionConstraint\\": \\"1.3.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/ghodss/yaml\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/go-redis/redis\\",
      \\"versionConstraint\\": \\"6.10.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/gogo/googleapis\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/gogo/protobuf\\",
      \\"versionConstraint\\": \\"^1.1.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/golang/protobuf\\",
      \\"versionConstraint\\": \\"1.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/golang/sync\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/google/go-github\\",
      \\"versionConstraint\\": \\"15.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/google/uuid\\",
      \\"versionConstraint\\": \\"0.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/googleapis/gax-go\\",
      \\"versionConstraint\\": \\"2.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/gorilla/mux\\",
      \\"versionConstraint\\": \\"1.6.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/gorilla/websocket\\",
      \\"versionConstraint\\": \\"1.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/grpc-ecosystem/go-grpc-middleware\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/grpc-ecosystem/go-grpc-prometheus\\",
      \\"versionConstraint\\": \\"1.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/grpc-ecosystem/grpc-opentracing\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/hashicorp/consul\\",
      \\"versionConstraint\\": \\"1.3.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/hashicorp/go-multierror\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/hashicorp/vault\\",
      \\"versionConstraint\\": \\"0.10.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/howeyc/fsnotify\\",
      \\"versionConstraint\\": \\"0.9.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/natefinch/lumberjack\\",
      \\"versionConstraint\\": \\"2.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/onsi/gomega\\",
      \\"versionConstraint\\": \\"1.3.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/open-policy-agent/opa\\",
      \\"versionConstraint\\": \\"0.8.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/openshift/api\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/opentracing/opentracing-go\\",
      \\"versionConstraint\\": \\"1.0.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/pmezard/go-difflib\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/prometheus/client_golang\\",
      \\"versionConstraint\\": \\"0.9.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/prometheus/client_model\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/prometheus/common\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/prometheus/prom2json\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/satori/go.uuid\\",
      \\"versionConstraint\\": \\"1.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/spf13/cobra\\",
      \\"versionConstraint\\": \\"0.0.2\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/stretchr/testify\\",
      \\"versionConstraint\\": \\"1.2.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/uber/jaeger-client-go\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"go.uber.org/atomic\\",
      \\"versionConstraint\\": \\"1.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"go.uber.org/multierr\\",
      \\"versionConstraint\\": \\"1.1.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"go.uber.org/zap\\",
      \\"versionConstraint\\": \\"1.7.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"golang.org/x/net\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"golang.org/x/oauth2\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"golang.org/x/time\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"golang.org/x/tools\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"google.golang.org/api\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"google.golang.org/genproto\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"google.golang.org/grpc\\",
      \\"versionConstraint\\": \\"^1.13.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"gopkg.in/yaml.v2\\",
      \\"versionConstraint\\": \\"2.2.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"istio.io/api\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/api\\",
      \\"versionConstraint\\": \\"release-1.11\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/apimachinery\\",
      \\"versionConstraint\\": \\"release-1.11\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/apiextensions-apiserver\\",
      \\"versionConstraint\\": \\"release-1.11\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/helm\\",
      \\"versionConstraint\\": \\"2.9.1\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/kubernetes\\",
      \\"versionConstraint\\": \\"release-1.11\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/gogo/status\\",
      \\"versionConstraint\\": \\"d60b5acac426cb52f7e28124c6aaa862d9f339f5\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/google/go-cmp\\",
      \\"versionConstraint\\": \\"0.2.0\\",
      \\"scopes\\": [
        \\"constraint\\"
      ]
    },
    {
      \\"name\\": \\"github.com/docker/distribution\\",
      \\"versionConstraint\\": \\"master\\",
      \\"scopes\\": [
        \\"override\\"
      ]
    },
    {
      \\"name\\": \\"k8s.io/client-go\\",
      \\"versionConstraint\\": \\"release-8.0\\",
      \\"scopes\\": [
        \\"override\\"
      ]
    },
    {
      \\"name\\": \\"github.com/Azure/go-autorest\\",
      \\"versionConstraint\\": \\"bca49d5b51a50dc5bb17bbf6204c711c6dbded06\\",
      \\"scopes\\": [
        \\"override\\"
      ]
    },
    {
      \\"name\\": \\"github.com/golang/glog\\",
      \\"scopes\\": [
        \\"override\\"
      ]
    },
    {
      \\"name\\": \\"github.com/imdario/mergo\\",
      \\"versionConstraint\\": \\"6633656539c1639d9d78127b7d47c622b5d7b6dc\\",
      \\"scopes\\": [
        \\"override\\"
      ]
    },
    {
      \\"name\\": \\"istio.io/istio/mixer/adapter/stackdriver/contextgraph\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"ignored\\"
      ]
    },
    {
      \\"name\\": \\"github.com/client9/misspell\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"required\\"
      ]
    },
    {
      \\"name\\": \\"github.com/istio/tools/protoc-gen-docs\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"required\\"
      ]
    },
    {
      \\"name\\": \\"fortio.org/fortio\\",
      \\"versionConstraint\\": \\"*\\",
      \\"scopes\\": [
        \\"required\\"
      ]
    }
  ]
}"
`;
